readme.md

# installation

```
npm install keysplitter
```


This tool lets you split up your crypto private key into 3 separate keys, each with
equal cryptographic strength.

Here's how you would use this:
1. Create your 3 split keys and store them in 3 different physical places
(preferably in a paper wallet or other offline storage)
2. Destroy the original private key. (yes, this is safer than keeping it)
3. Later, when you need to access your crypto account again, you can restore your
private key with any 2 out of the 3 split keys.

Why is this better than pretty much all other methods out there?
- storing your private keys on your computer makes it susceptible to theft by a hacker.
- if you encrypt your private keys using some password, you now need to worry about finding
a safe place to keep your password. If you don't write it down anywhere, and you forget it,
or something was to happen to you, you and/or your family will lose it all.
- if you store your private key on a paper wallet in your safe, you can lose everything
through theft or if your house burns down.
- if you keep your private key in multiple places (for redundancy), it will make it
less likely that you will lose it, but it will also make it more likely to get stolen.
- if you cut up your private key into 2 or more pieces and store it in multiple places,
you can lose access to all your crypto funds if you lose just one piece.


The split keys that this script generates eliminates all of the above risks in an elegant
and ultra-secure way.

If any of your 3 keys get stolen, nothing is lost. It's useless to a thief, and you can
still access your account using the other 2 keys.

That said, use this tool at your own risk. I accept no responsibility if you lose your private key or crypto while using this tool.
Always make sure your system isn't compromised. Also test restoring your private key from the split keys before destroying the original private key.

## using this library


```javascript
const assert = require('assert');
const Splitter = require('splitter');
const splitter = new Splitter();

const originalKey = '3a1076bf45ab87712ad64ccb3b10217737f7faacbf2872e88fdd9a537d8fe266';
let keys = splitter.splitPrivateKey(originalKey);
console.dir(keys);
//[ '010b5701057440e7740c9628a00ba62a63ea727b04871041c40b1168e6b2a5963e158dd64c1588c8714efcf8885ebc50',
//  '020bda53f693d2a1c9ef2169640a6380696f3fc5439a6fb00b6d1173cb054cf37dbcfeec1588c8714efcf8885ebc50',
//  '030b2f98bece0b577ff9746a1c0a676047acd4af8507bcda0b6d1173cb054cf37dbcfeec1168e6b2a5963e158dd64c' ]

// We can restore the private key with split key 0 and key 1
assert(splitter.restorePrivateKey(keys[0], keys[1]) === originalKey);  // true

// or we can restore it using key 0 and key 2
assert(splitter.restorePrivateKey(keys[0], keys[2]) === originalKey);  // true

// or we can restore it using key 1 and key 2
assert(splitter.restorePrivateKey(keys[1], keys[2]) === originalKey);  // true
```

## Command line tool

To create split keys:

```
keysplitter split 3a1076bf45ab87712ad64ccb3b10217737f7faacbf2872e88fdd9a537d8fe266
```

(add the -b or --barcode option to have it also generate barcode PNG images)

To restore private key from any 2 split keys:

```
keysplitter restore 120bc4f5f571d3b89ececb47770a7d266a8d3edc14bb859e0b1b9e7951556ca8aa51958e0ba8e574e2d501c562f6ed \
020b31b893cba77e86b44820a10a79406aa978867c4a80900b1b9e7951556ca8aa51958e0fcee550a48f693467f85f
```

A tool to restore a private key directly from barcode images is in the works.


changelog.md

## changelog

**0.2.0** initial release

**0.3.0** now using random numbers instead of part of a split key for
barcode filename generation. Otherwise, one out of the 3 keys would
be slightly less secure.

**0.4.0** Updated version number in binary to be consistent with
package.json


test.js
const Splitter = require('./splitter');


const key = ;

const splitter = new Splitter();


let codes = splitter.splitPrivateKey(key);

console.log("generated codes: ");
for (k of codes) {
  console.log(k.toString('base64'));
}
for (k of codes) {
  console.log(k.toString('hex'));
}


test/splitter_test.js
const assert = require('assert');
const should = require('chai').should();
const expect = require('chai').expect;

const Splitter = require('../src/splitter');

let splitter = new Splitter();
let testKey = '3a1076bf45ab87712ad64ccb3b10217737f7faacbf2872e88fdd9a537d8fe266';


// Convert a byte array to a hex string
function toHex(bytes) {
  return splitter.toHexStr(bytes);
}
// Convert a hex string to a byte array
function hexToBytes(hex) {
  return splitter.toBytes(hex);
}


describe('splitter', () => {

    describe("splitInto3Parts", () => {
        it("should split the original key in 3 parts", () => {
            let res = splitter.splitInto3Segments(testKey);
            res.keySegments.should.have.lengthOf(3);
        });

        it("should return a key length of 11", () => {
            let res = splitter.splitInto3Segments(testKey);
            res.keyLength.should.equal(11);
        });

        it("should add up to total the length of the original key", () => {
            let res = splitter.splitInto3Segments(testKey);
            let keys = res.keySegments;
            let total = 0;
            for (let i of keys) {
                total += i.length;
            }
            total.should.equal(32);
        })

        it("should work with short keys as well", () => {
            let res = splitter.splitInto3Segments("11223344556677889900");
            res.keyLength.should.equal(4);
            toHex(res.keySegments[0]).should.equal("112233");
            toHex(res.keySegments[1]).should.equal("445566");
            toHex(res.keySegments[2]).should.equal("77889900");
        });

    });

    describe("makeOTPKeys", () => {
        it("should always create 3 keys", () => {
            let keys = splitter.createOTPKeys(11);
            keys.should.have.lengthOf(3);
        });
        it("should create keys of the specified length", () => {
            let keys = splitter.createOTPKeys(11);
            keys[0].length.should.equal(11);
            keys[1].length.should.equal(11);
            keys[2].length.should.equal(11);
        });
    });

    describe("makeSplitKey", () => {
        beforeEach(() => {
            let res = splitter.splitInto3Segments(testKey);
            this.keyLength = res.keyLength;
            this.keySegments = res.keySegments;
            this.pads = splitter.createOTPKeys(res.keyLength);
        });

        it("should generate type 1 key", () => {
            let key1 = splitter.makeSplitKey(0x01, this.keySegments, this.keyLength, this.pads);
            key1[0].should.equal(0x01);
            key1.should.have.lengthOf(48);
        });
        it("should generate type 2 key", () => {
            let key1 = splitter.makeSplitKey(0x02, this.keySegments, this.keyLength, this.pads);
            key1[0].should.equal(0x02);
            key1.should.have.lengthOf(47);
        });
        it("should generate type 3 key", () => {
            let key1 = splitter.makeSplitKey(0x03, this.keySegments, this.keyLength, this.pads);
            key1[0].should.equal(0x03);
            key1.should.have.lengthOf(47);
        });
        // TODO: we could test more values in the buffer here but those are implictly tested in the
        // end-to-end test.
    });

    describe("splitPrivateKey", () => {
        it ("should generate three split keys", () => {
            let splitKeys = splitter.splitPrivateKey(testKey);
            splitKeys.should.have.lengthOf(3);
        });
        it ("key 1 should be 48 bytes long", () => {
            let splitKeys = splitter.splitPrivateKey(testKey);
            splitKeys[0].should.have.lengthOf(48*2);
        });
        it ("key 2 should be 47 bytes long", () => {
            let splitKeys = splitter.splitPrivateKey(testKey);
            splitKeys[1].should.have.lengthOf(47*2);
        });
        it ("key 3 should be 47 bytes long", () => {
            let splitKeys = splitter.splitPrivateKey(testKey);
            splitKeys[2].should.have.lengthOf(47*2);
        });
    });

    describe("extractCode", () => {
        it ("should extract the right segments from a split key", () => {
            let splitKey = "010b2fc62109187bbb70f391e50bdeed47972ae7cbfb75f8810b87a892ec1b1d855a793e05163914ecdc4174541ff44a";
            let key = hexToBytes(splitKey);
            let p = splitter.extractCode(key);

            p.partType.should.equal(0x01);
            toHex(p.part1).should.equal("2fc62109187bbb70f391e5");
            toHex(p.part2).should.equal("deed47972ae7cbfb75f881");
        });
        it ("should extract the right OTP keys from a split key", () => {
            let splitKey = "010b2fc62109187bbb70f391e50bdeed47972ae7cbfb75f8810b87a892ec1b1d855a793e05163914ecdc4174541ff44a";
            let key = hexToBytes(splitKey);
            let p = splitter.extractCode(key);

            p.partType.should.equal(0x01);
            toHex(p.pad1).should.equal("87a892ec1b1d855a793e05");
            toHex(p.pad2).should.equal("163914ecdc4174541ff44a");
        });
    });

    describe("retrievePads", () => {
        const splitKey1 = "010bd95d278679faee51d17d0a0b287641184b669eda57146e0b0c9e3f5766f8ee23ca3c7bcca3c42e9b028b50ab9618";
        const splitKey2 = "020bc7a52f7611cf19d96683530a7e76b08afcab93ac285a0be34d51393c516920fbab46cca3c42e9b028b50ab9618";
        const splitKey3 = "030bf6b3b291dea90c218140540abe4b4bf30151f6df49f00be34d51393c516920fbab460c9e3f5766f8ee23ca3c7b";

        it ("should succesfully extract all OTP keys from 2 split keys", () => {
            let key1 = hexToBytes(splitKey1);
            let key2 = hexToBytes(splitKey2);
            let p1 = splitter.extractCode(key1);
            let p2 = splitter.extractCode(key2);

            let pads = splitter.retrievePads(p1, p2);
            toHex(pads[0]).should.equal("e34d51393c516920fbab46");
            toHex(pads[1]).should.equal("0c9e3f5766f8ee23ca3c7b");
            toHex(pads[2]).should.equal("cca3c42e9b028b50ab9618");
        });

        it ("should throw an exception if pads don't match", () => {
            let sk1 = splitKey1.substr(0,splitKey1.length -1) + 'a'; // manipulate one of the keys
            let key1 = new Buffer(sk1, 'hex');
            let key2 = new Buffer(splitKey2, 'hex');
            let p1 = splitter.extractCode(key1);
            let p2 = splitter.extractCode(key2);

            let badFn = () => {
                splitter.retrievePads(p1, p2);
            }
            expect(badFn).to.throw(Error, /pad 2 exists in both codes but do not match/);
        });
    });

    describe("encode and decode", () => {
        beforeEach(() => {
            this.splitKeys = splitter.splitPrivateKey(testKey);
        });

        it("should successfully decode with keys 1 and 2", () => {
            let fullKey = splitter.restorePrivateKey(this.splitKeys[0], this.splitKeys[1]);
            fullKey.should.equal(testKey);
        });
        it("should successfully decode with keys 2 and 1", () => {
            let fullKey = splitter.restorePrivateKey(this.splitKeys[1], this.splitKeys[0]);
            fullKey.should.equal(testKey);
        });
        it("should successfully decode with keys 1 and 3", () => {
            let fullKey = splitter.restorePrivateKey(this.splitKeys[0], this.splitKeys[2]);
            fullKey.should.equal(testKey);
        });
        it("should successfully decode with keys 3 and 1", () => {
            let fullKey = splitter.restorePrivateKey(this.splitKeys[2], this.splitKeys[0]);
            fullKey.should.equal(testKey);
        });
        it("should successfully decode with keys 2 and 3", () => {
            let fullKey = splitter.restorePrivateKey(this.splitKeys[1], this.splitKeys[2]);
            fullKey.should.equal(testKey);
        });
        it("should successfully decode with keys 3 and 2", () => {
            let fullKey = splitter.restorePrivateKey(this.splitKeys[2], this.splitKeys[1]);
            fullKey.should.equal(testKey);
        });
    });
});


crypt.js


const sha1 = require('sha1');
const QRCode = require('qrcode');


const crypto = require('crypto'),
  algorithm = 'aes-256-ctr';

//const IV = crypto.randomBytes(16);
const IV = Buffer.alloc(16);
console.log("Using IV: ");
console.log(IV);

// currently support splitting the key in 3 sections.
const NUM_PARTS = 3;
// will be encoded in the barcode so we know which of the 3 barcode types we're dealing with
const PART_COMBINATIONS = [
  0x12,  // part 1 and 2
  0x23,  // part 2 and 3
  0x13,  // part 1 and 3
];

//const IV = Buffer.from(crypto.randomBytes(16)'someIsdsfdsdfkhkj28fskdfsdfdsV'.slice(0,16), 'hex');

function secretKey(password) {
  let pwd = sha1(password).substr(0,32);
  //console.log("using password: " + pwd);
  return new Buffer(pwd);
}

/**
 * encrypts a given Buffer and returns a base64 string
 * @param buf
 * @param pwd
 * @returns {*}
 */
function encrypt(buf, pwd) {
  let cipher = crypto.createCipheriv(algorithm, secretKey(pwd), IV);
  let crypted = cipher.update(buf, null, 'base64');
  return crypted + cipher.final('base64');
}

function decrypt(text, pwd){
  let decipher = crypto.createDecipheriv(algorithm, secretKey(pwd), IV);
  let dec = decipher.update(text, 'base64')
  dec += decipher.final();
  return dec;
}

function genBarcodeData(key1, key2, key3) {
  let k1 = key1 ? (new Buffer(key1).toString('hex')) : 0;
  let k2 = key2 ? (new Buffer(key2).toString('hex')) : 0;
  let k3 = key3 ? (new Buffer(key3).toString('hex')) : 0;
  return `${k1}|${k2}|${k3}`;
}

/**
 * creates the QR code string. This is assembled as follows:
 *
 * 1 byte: part combination (one of the PART_COMBINATIONS values)
 * 1 byte: length of the first key part
 * n bytes: the first key part
 * 1 byte: lenght of the second key part
 * m bytes: the second key part
 * @param partCombo
 * @param parts
 * @param pwd
 * @returns buffer containing the barcode data
 */
function genEncryptedBarcodeData(partCombo, parts, pwd) {
  if (pwd) {
    // we are going to put the parts together in a binary buffer,
    // before encrypting the key
  } else {
    // we are going to embed the key in plain text, so it can
    // be reconstructed by reading it with any barcode reader and putting
    // the parts back together
    if (partCombo === PART_COMBINATIONS[0]) {
      // part 1 and 2
      key1 = parts[0];
      key2 = parts[1];
      key3 = 0;
    } else if (partCombo === PART_COMBINATIONS[1]) {
      // part 1 and 2
      key1 = 0;
      key2 = parts[0];
      key3 = parts[1];
    } else if (partCombo === PART_COMBINATIONS[3]) {
      // part 1 and 3
      key1 = 0;
      key2 = parts[0];
      key3 = parts[1];
    }
  }


  let buf = [partCombo];
  parts.forEach((part) => {
    buf.push(part.length);
    buf.push(...part);
  });

  console.log("part " + partCombo.toString(16));
  console.dir((new Buffer(buf)).toString('hex'));

  buf = new Buffer(buf);
  if (pwd) {
    console.log("need to encrypt");
    let enc = encrypt(buf, pwd);
    console.log(buf.toString('hex') + " encrypted = " + enc);
    return enc; // already base64 encoded
  } else {
    // add the '-' to indicate the key is not encrypted with a password.
    return (new Buffer(buf)).toString('base64') + '-';
  }
}

/**
 * generates all 3 barcodes
 * @param key
 * @param password
 * @returns array of QR barcodes
 */
function splitAndEncrypt(key, password) {
  // read key into binary buffer
  const buf = new Buffer(key, 'hex');
  // convert to array
  const arr = [...buf];

  // we're going to split in 3 parts
  let start=0, end = 0;
  let bufpart = [];
  let partLen = Math.round(buf.length / NUM_PARTS);
  console.log(" part length = " + partLen);
  for (let i=0; i<NUM_PARTS; i++) {
    if (i == NUM_PARTS - 1) {
      // last element store the remainder
      end = buf.length;
    } else {
      end = start + partLen;
    }
    console.log(`adding ${start} to ${end}`);
    bufpart[i] = buf.slice(start, end);
    start = start + bufpart[i].length;
    console.log(`part ${i}: ${(new Buffer(bufpart[i])).toString('hex')}`);
  }

  // now that we've split the key, create the 3 bar codes
  let barcodes = [
    genBarcodeData(PART_COMBINATIONS[0], [bufpart[0], bufpart[1]], pwd),
    genBarcodeData(PART_COMBINATIONS[1], [bufpart[1], bufpart[2]], pwd),
    genBarcodeData(PART_COMBINATIONS[2], [bufpart[0], bufpart[2]], pwd)
  ];
  return barcodes;
}

function extractKeyParts(str) {
  console.log("extracted key: ");
  //console.log((new Buffer(str, 'base64')))

  // move the data into a byte array
  let arr = [...(new Buffer(str, 'base64'))];
  let partType = arr[0]; // should be one of the PART_COMBINATIONS values
  let keypart = [];
  let c = 1;
  let i = 0;
  do {
    let length = arr[c];
    let buf = arr.slice(c+1, c+1+length);
    keypart[i] = (new Buffer(buf)).toString('hex')
    c += length + 1;
    i += 1;
  } while (c < arr.length)

  // put the key parts back into a 3 element array, based on which type (PART_COMBINATIONS)
  // this key was.
  let keys;
  switch (partType) {
    case 0x12:
      keys = [keypart[0], keypart[1], null];
      break;
    case 0x23:
      keys = [null, keypart[0], keypart[1]];
      break;
    case 0x13:
      keys = [keypart[0], null, keypart[1]];
      break;
    default:
      throw new Error("unknown key part: " + partType);
  }
  // console.log("extracted");
  // console.dir(keys);
  return keys;
}

/**
 * restores the original key from 2 QRcode values, and an optional password.
 * @param key1
 * @param key2
 * @param pwd
 * @returns {string}
 */
function decryptWithKeys(key1, key2, pwd) {
  let dec1, dec2;
  if (key1.endsWith('-') && key2.endsWith('-')) {
    console.log("This QR Code split key was not password protected")
    dec1 = extractKeyParts(key1.substr(0,key1.length-1));
    dec2 = extractKeyParts(key2.substr(0,key2.length-1));
  } else if (!key1.endsWith('-') && !key2.endsWith('-')) {
    console.log("This QR Code split key was password protected");
  } else {
    throw new Error("These barcodes are not for the same key");
  }

  let d = [];
  for (let i=0; i<3; i++) {
    // either dec1 or dec2 should contain each part
    if (dec1[i] && dec2[i] && dec1[i] != dec2[i]) {
      throw new Error(`key 1 and key 2 are not associated with the same private key!`);
    }
    d[i] = dec1[i] ? dec1[i] : dec2[i];
  }
  return d.join('');
}


/**
 * generates 3 QR codes from an encrypted private key. This private key can be decrypted
 * again with the password, and with 2 of the QR codes.
 * @param key array of data to put in QR Code
 * @param filename filename to store QR Code at
 */
function generateQrCode(str, filename, callback) {
  let options = {
    color: {
      dark: '#00F',  // Blue dots
      light: '#0000' // Transparent background
    },
    errorCorrectionLevel: 'M'
  };

  QRCode.toFile(filename, str, options, (err) => {
    if (err) throw err;
    if (callback) callback();
  });
}


const key = '3a1076bf45ab87712ad64ccb3b10217737f7faacbf2872e88fdd9a537d8fe266';
//const pwd = 'anothrd';
const pwd = null;
//onst pwd = '';




// first get the 3 encrypted keys
let barcodes = splitAndEncrypt(key, pwd);
// then create the QR codes
barcodes.forEach((buf, i) => {
  console.log("barcode as base64 " + buf.toString('base64'));
  generateQrCode(buf, `pk-${i}.png`);
});

//
console.log("original key");
console.log(key);
//
//
let fullKey = decryptWithKeys(barcodes[0], barcodes[1], pwd);
console.log("constructed from 0 and 1");
console.log(fullKey);

fullKey = decryptWithKeys(barcodes[1], barcodes[2], pwd);
console.log("constructed from 1 and 2");
console.log(fullKey);

fullKey = decryptWithKeys(barcodes[0], barcodes[2], pwd);
console.log("constructed from 0 and 2");
console.log(fullKey);


src/splitkey.js


const program = require('commander');
const Splitter = require('./splitter');
const QRCode = require('qrcode');
const splitter = new Splitter();

function generateQrCode(str, filename, callback) {
    let options = {
        color: {
            dark: '#000',  // Blue dots
            light: '#0000' // Transparent background
        },
        errorCorrectionLevel: 'L'
    };

    QRCode.toFile(filename, str, options, (err) => {
        if (err) throw err;
        if (callback) callback();
    });
}


program
  .version('1.0.0')

// configure the split command
program
  .command('split <key>')
  .description('split private key into 3 redundant split keys')
  .option('-b, --barcode', 'generate barcodes')
  .action((k, options) => {
    console.log("splitting key " + k);

    let uid;
    splitter.splitPrivateKey(k).forEach((key, c) => {
      let str = key.toString('hex');
      // the unique filename is based on the last 4 characters of split key 1
      if (!uid) {
        uid = str.slice(-4);
      }
      console.log(`\nkey ${c+1}:\n${str}`);
      if (options.barcode) {
        // generate barcode as well
        generateQrCode(str, `sk-${uid}-${c}.png`);
      }
    });

    if (options.barcode)
      console.log('\ngenerated barcodes.');
  });

program
  .command('restore <key1> <key2>')
  .description('restore private key from 2 split keys')
  .action((k1, k2) => {
    let restoredKey = splitter.restorePrivateKey(k1, k2);
    console.log(`\nRestored key:\n${restoredKey}`);
  });

program.parse(process.argv);


examples/encode-decode.js
const assert = require('assert');
const Splitter = require('../src/splitter');

splitter = new Splitter();

originalKey = '3a1076bf45ab87712ad64ccb3b10217737f7faacbf2872e88fdd9a537d8fe266';
keys = splitter.splitPrivateKey(originalKey);
console.dir(keys);
// 010b81ea86f6207d633700331c0b70c1e06812e113bc865a780b9f84d1c756a4a892e438692421d9336e7e80e31e9dcd
// 120b54bfc1e621935f684887410aed6c5e1accf7d51d065e0bbbfaf04965d6e4462ae5502421d9336e7e80e31e9dcd
// 020b1e31af8c2bd50792344b810a56c956eef42dfd6cfcfb0bbbfaf04965d6e4462ae5509f84d1c756a4a892e43869

// restore private key with split key 0 and key 1
assert(splitter.restorePrivateKey(keys[0], keys[1]) === originalKey);
// true

// restore private key with split key 0 and key 2
assert(splitter.restorePrivateKey(keys[0], keys[2]) === originalKey);
// true

// restore private key with split key 1 and key 2
assert(splitter.restorePrivateKey(keys[1], keys[2]) === originalKey);


src/splitter.js

// represents the 3 different types of codes that are generated. Each starts out with one of these values,
// which indicate with private key segments and which pad keys it embeds.
const PART_COMBINATIONS = [
    0x01,  // part 0 and 1
    0x02,  // part 1 and 2
    0x03   // part 0 and 2
];
// this will be incremented if the split method is changed
const SPLIT_METHOD_VERSION = 0;

const Splitter = function() {

  // define getRandomBytes so it works in Node as well as in browsers.
  const getRandomBytes = (
    (typeof self !== 'undefined' && (self.crypto || self.msCrypto))
      ? function() { // Browsers
          var crypto = (self.crypto || self.msCrypto), QUOTA = 65536;
          return function(n) {
            var a = new Uint8Array(n);
            for (var i = 0; i < n; i += QUOTA) {
              crypto.getRandomValues(a.subarray(i, i + Math.min(n - i, QUOTA)));
            }
            return a;
          };
        }
      : function() { // Node
          return require("crypto").randomBytes;
        }
  )();


  function decodeKey(keyparts, res, pads) {
      let parts = [];
      let part_combination = (res.partType & 0x0F);
      if (part_combination === PART_COMBINATIONS[0]) {
          // key type 1, contains part 1 and 2, encoded with pad key 1
          parts[0] = xor(res.part1, pads[0]);
          parts[1] = xor(res.part2, pads[0]);
      } else if (part_combination === PART_COMBINATIONS[1]) {
          // key type 2, contains part 2 and 3, encoded with pad key 2
          parts[1] = xor(res.part1, pads[1]);
          parts[2] = xor(res.part2, pads[1]);
      } else if (part_combination === PART_COMBINATIONS[2]) {
          // key type 3, contains part 1 and 3, encoded with pad key 3
          parts[0] = xor(res.part1, pads[2]);
          parts[2] = xor(res.part2, pads[2]);
      }
      // now compare any parts we found with the keyparts passed in. If the keyparts array already
      // contains elements they should match with what we found here
      for (let i = 0; i < 3; i++) {
          if (parts[i]) {
              if (keyparts[i] && bytesToHex(keyparts[i]) !== bytesToHex(parts[i])) {
                  throw new Error(`decrypted key parts in both codes do not match! ${bytesToHex(keyparts[i])} / ${bytesToHex(parts[i])}`);
              }
              keyparts[i] = parts[i];
          }
      }
  }

  // Convert a hex string to a byte array
  function hexToBytes(hex) {
      let bytes = new Uint8Array(hex.length/2);
      for (c = 0; c < hex.length/2; c += 1)
      bytes[c] = parseInt(hex.substr(c*2, 2), 16);
      return bytes;
  }

  // Convert a byte array to a hex string
  function bytesToHex(bytes) {
      for (var hex = [], i = 0; i < bytes.length; i++) {
          hex.push((bytes[i] >>> 4).toString(16));
          hex.push((bytes[i] & 0xF).toString(16));
      }
      return hex.join("");
  }

  /**
   * performs an xor operation on bytes of a given array. It will always return an array
   * with the same length of a, even if b (the xor filter) is longer
   *
   * @param a key to perform xor operation on
   * @param b the xor filter (one-time-pad)
   * @returns array
   */
  function xor(a, b) {
      return a.map((c, i) => {
          return a[i] ^ b[i];
      });
  }

  /**
   * for testing only, prints the object that extractCode() returns.
   * @param code
   */
  function printCode(code) {
      console.log("\nkeyType: " + code.partType);
      console.log("part 1: " + bytesToHex(code.part1));
      console.log("part 2: " + bytesToHex(code.part2));
      console.log("pad 1: " + bytesToHex(code.pad1));
      console.log("pad 2: " + bytesToHex(code.pad2));
  }


return {

    /**
     * Splits a private key (ie Ethereum or Bitcoin private key) and generate 3 separate codes out of them.
     * @param key private key (as a hex string)
     * @returns an array of 3 new keys, each as strong as the original one.
     */
    splitPrivateKey(key) {
        let {keySegments, keyLength} = this.splitInto3Segments(key);

        // generate 3 random One Time Pad keys, at the length of one third of the private key length
        let pads = this.createOTPKeys(keyLength);

        keys = [
            this.makeSplitKey(PART_COMBINATIONS[0], keySegments, keyLength, pads),
            this.makeSplitKey(PART_COMBINATIONS[1], keySegments, keyLength, pads),
            this.makeSplitKey(PART_COMBINATIONS[2], keySegments, keyLength, pads)
        ];
        return keys.map((k) => { return bytesToHex(k); });
    },

    toBytes(hexStr) {
      return hexToBytes(hexStr);
    },

    toHexStr(bytes) {
      return bytesToHex(bytes);
    },

    /**
     * restores the original private key using any 2 of the 3 keys generated by splitPrivateKey()
     * @param splitKey1 a buffer containing one of the split keys
     * @param splitKey2 a buffer containing one other split key
     * @returns the original private key
     */
    restorePrivateKey(splitKey1, splitKey2) {
        // convert to byte array if passed in as String
        if (typeof splitKey1 === 'string') {
          splitKey1 = hexToBytes(splitKey1);
          splitKey2 = hexToBytes(splitKey2);
        }

        let res1 = this.extractCode(splitKey1);
        //printCode(res1);

        let res2 = this.extractCode(splitKey2);
        //printCode(res2);

        let pads = this.retrievePads(res1, res2);

        let keySegments = [];
        decodeKey(keySegments, res1, pads);
        // console.log("after parting code 1:");
        // if (keyparts[0]) console.log("part 1: " + asHex(keyparts[0]));
        // if (keyparts[1]) console.log("part 2: " + asHex(keyparts[1]));
        // if (keyparts[2]) console.log("part 3: " + asHex(keyparts[2]));

        decodeKey(keySegments, res2, pads);
        // console.log("after parting code 2:");
        // if (keyparts[0]) console.log("part 1: " + asHex(keyparts[0]));
        // if (keyparts[1]) console.log("part 2: " + asHex(keyparts[1]));
        // if (keyparts[2]) console.log("part 3: " + asHex(keyparts[2]));

        return bytesToHex(keySegments[0]) + bytesToHex(keySegments[1]) + bytesToHex(keySegments[2]);
    },


    /***************************************************************************************************
     * all methods below should be considered private, but are exported anyway to make this class
     * better testable.
     ***************************************************************************************************/

    /**
     * generates 3 OTP keys of equal length, which are used to encrypt the key segments for each key type
     * @param length key length to use (typically 11, as 3x11 = 33 bytes, which fits an Ethereum private key)
     * @returns array of OTP keys (of type Buffer)
     */
    createOTPKeys(length) {
        let pads = [];
        for (let i = 0; i < 3; i++) {
            pads[i] = getRandomBytes(length);
            // console.log("created secret key " + asHex(pads[i]));
        }
        return pads;
    },

    /**
     * splits the private key into three seqments (almost) equal in size. For 32 byte Ethereum private keys, this will
     * return 3 buffers containing 11, 11 and 10 bytes.
     * @param key the private key (string) to split up
     */
    splitInto3Segments(key) {
        const bytes = hexToBytes(key);

        // we're going to split in 3 parts
        let start = 0, end = 0;
        let bufpart = [];
        let partLen = Math.round(bytes.length / 3);
        for (let i = 0; i < 3; i++) {
            if (i === 2) {
                // last element store the remainder
                end = bytes.length;
            } else {
                end = start + partLen;
            }
            bufpart[i] = bytes.slice(start, end);
            start = start + bufpart[i].length;
        }

        let keyLength = Math.max(partLen, bufpart[2].length); // this will typically be 11 for Ethereum private keys
        return {keySegments: bufpart, keyLength};
    },

    /**
     * constructs one of the three split keys, consisting of 2/3s of the key (encoded using OTP) and two pad keys.
     * The OTP key that the private key segment is encoded with, only exists in the other keys.
     * The keys are encoded as follows (using example of a 32 byte Ethereum private key)
     *
     * byte 1:      part type identifier (one of the PART_COMBINATIONS values)
     * byte 2:      length of the first key segment (encoded with OTP key)
     * byte 3-13:   first key segment
     * byte 14:     length of the second key segment (encoded with OTP key)
     * byte 15-25:  second key segment
     * byte 26:     length of the OTP keys used (11 for ethereum keys)
     * byte 27-37:  first OTP key
     * byte 38-48:  second OTP key
     *
     * @param partyType
     * @param keySegments  array contain all 3 key segments
     * @param keyLength    OTP key length (typically 11)
     * @param pads         array of 3 (11 byte) random OTP keys
     */
    makeSplitKey(partType, keySegments, keyLength, pads) {
        // generate 3 random keys to be used
        let buf = [(SPLIT_METHOD_VERSION << 4) + partType];
        let p1, p2, keys;
        if (partType === PART_COMBINATIONS[0]) {
            p1 = xor(keySegments[0], pads[0]);
            p2 = xor(keySegments[1], pads[0]);
            keys = [pads[1], pads[2]];   // add the other two ciphers, one of which will be needed to decrypt
        } else if (partType === PART_COMBINATIONS[1]) {
            p1 = xor(keySegments[1], pads[1]);
            p2 = xor(keySegments[2], pads[1]);
            keys = [pads[0], pads[2]];
        } else if (partType === PART_COMBINATIONS[2]) {
            p1 = xor(keySegments[0], pads[2]);
            p2 = xor(keySegments[2], pads[2]);
            keys = [pads[0], pads[1]];
        }
        buf.push(p1.length);
        buf.push(...p1);
        buf.push(p2.length);
        buf.push(...p2);
        buf.push(keyLength);
        buf.push(...keys[0]);
        buf.push(...keys[1]);

        return new Uint8Array(buf);
    },

    /**
     * accepts a buffer or array with a split key, and extracts the key segments and OTP keys from it
     * @param str
     * @returns object containing partType, part1, part2, pad1 and pad2 properties
     */
    extractCode(str) {
        //console.log((new Buffer(str, 'base64')))
        let res = {}, length;
        // move the data into a byte array
        let arr = str;
        res.partType = arr[0]; // should be one of the PART_COMBINATIONS values
        length = arr[1]; // length of the first key part
        let c = 2;
        res.part1 = arr.slice(c, c + length);
        c += length;
        length = arr[c]; // length of the second key part
        c += 1;
        res.part2 = arr.slice(c, c + length);
        let keylength = arr[c + length];
        c += length + 1;
        res.pad1 = arr.slice(c, c + keylength);
        c += keylength;
        res.pad2 = arr.slice(c, c + keylength);
        return res;
    },


    /**
     * retrieves the OTP keys from both split keys. There's 3 of them, but each split key only holds
     * 2. This function with return an array of pad key 1,2, and 3 in the right order.
     *
     *  a(key0)  b(key0)  key1  key2   TYPE 1
     *  b(key1)  c(key1)  key0  key2   TYPE 2
     *  a(key2)  c(key2)  key0  key1   TYPE 3
     */
    retrievePads(res1, res2) {
        let pads = [[], [], []];
        for (let res of [res1, res2]) {
            if (res.partType === PART_COMBINATIONS[0]) {
                // type 1 code contains pad 2 and pad 3
                pads[1].push(res.pad1);
                pads[2].push(res.pad2);
            } else if (res.partType === PART_COMBINATIONS[1]) {
                // type 2 code contains pad 1 and pad 3
                pads[0].push(res.pad1);
                pads[2].push(res.pad2);
            } else if (res.partType === PART_COMBINATIONS[2]) {
                // type 3 code contains pad 1 and pad 2
                pads[0].push(res.pad1);
                pads[1].push(res.pad2);
            }
        }

        // check that if we extracted the same key multiple times, they must match. If not,
        // one of the codes was read incorrectly or tampered with
        let p = pads.map((pad, i) => {
            if (pad.length === 0) {
                throw new Error(`no pad key data for pad ${i}`);
            } else if (pad.length === 2) {
                let p1 = bytesToHex(pad[0]);
                let p2 = bytesToHex(pad[1]);
                if (p1 !== p2) {
                    throw new Error(`pad ${i} exists in both codes but do not match! ${p1} / ${p2}`);
                }
            }
            return pad[0];
        });
        return p;
    }
  }
}




// export this module
if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
    module.exports = Splitter;
} else {
    if (typeof define === 'function' && define.amd) {
        define([], function() {
            return Splitter;
        });
    } else {
        window.Splitter = Splitter;
    }
}
